<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Practical Skills for Biology Research II</title>
    <meta charset="utf-8" />
    <meta name="author" content="Niklas Mähler" />
    <meta name="date" content="2021-10-04" />
    <script src="libs/header-attrs-2.8/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/niklas.css" type="text/css" />
    <link rel="stylesheet" href="css/niklas-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: left, middle, inverse, title-slide

# Practical Skills for Biology Research II
## R packages
### Niklas Mähler
### Umeå University
### 2021-10-04

---




# Learning goals

- Can describe what an R package is
- Know where to search for existing R packages
- Install packages from different sources
- See what packages are loaded in a session
- Basic knowledge of namespaces

---

# R packages

.left-column[
## What are they?
]

.right-column[
Source code that someone else wrote.

Collection of functions and/or data that collectively are aimed at solving a particular problem.


|Package  |Description                           |
|:--------|:-------------------------------------|
|ggplot2  |Data visualisation                    |
|pheatmap |Plotting pretty heatmaps              |
|readxl   |Reading Excel sheets into R           |
|dplyr    |Data frame manipulation               |
|DESeq2   |Differential gene expression analysis |

]

---

# R packages

.left-column[
## What are they?
## How to use them
]

.right-column[
The `library` function loads the package and makes anything exported from the package directly available.
Whatever is exported from the package will be available for the duration of the session.


```r
library(ggplot2)
head(luv_colours)
```

```
##          L             u         v           col
## 1 9341.570 -3.370649e-12    0.0000         white
## 2 9100.962 -4.749170e+02 -635.3502     aliceblue
## 3 8809.518  1.008865e+03 1668.0042  antiquewhite
## 4 8935.225  1.065698e+03 1674.5948 antiquewhite1
## 5 8452.499  1.014911e+03 1609.5923 antiquewhite2
## 6 7498.378  9.029892e+02 1401.7026 antiquewhite3
```

&gt; "Unpack the box and put all the items on the table"
]

---

# R packages

.left-column[
## What are they?
## How to use them
]

.right-column[
Another way of using them:


```r
head(ggplot2::luv_colours)
```

```
##          L             u         v           col
## 1 9341.570 -3.370649e-12    0.0000         white
## 2 9100.962 -4.749170e+02 -635.3502     aliceblue
## 3 8809.518  1.008865e+03 1668.0042  antiquewhite
## 4 8935.225  1.065698e+03 1674.5948 antiquewhite1
## 5 8452.499  1.014911e+03 1609.5923 antiquewhite2
## 6 7498.378  9.029892e+02 1401.7026 antiquewhite3
```

&gt; "Only take a specific item out of the box, use it, then put it back when you're done with it"
]

---

# R packages

.left-column[
## What are they?
## How to use them
## What is loaded?
]

.right-column[

```r
sessionInfo()
```

```
## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] here_1.0.1        htmltools_0.5.1.1 forcats_0.5.1     stringr_1.4.0    
##  [5] dplyr_1.0.6       purrr_0.3.4       readr_1.4.0       tidyr_1.1.3      
##  [9] tibble_3.1.2      ggplot2_3.3.3     tidyverse_1.3.1   conflicted_1.0.4 
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.1.1 xfun_0.23        haven_2.4.1      colorspace_2.0-1
##  [5] vctrs_0.3.8      generics_0.1.0   yaml_2.2.1       utf8_1.2.1      
##  [9] rlang_0.4.11     pillar_1.6.1     glue_1.4.2       withr_2.4.2     
## [13] DBI_1.1.1        dbplyr_2.1.1     modelr_0.1.8     readxl_1.3.1    
## [17] lifecycle_1.0.0  munsell_0.5.0    gtable_0.3.0     cellranger_1.1.0
## [21] rvest_1.0.0      evaluate_0.14    knitr_1.33       fastmap_1.1.0   
## [25] fansi_0.5.0      broom_0.7.6      Rcpp_1.0.6       backports_1.2.1 
## [29] scales_1.1.1     cachem_1.0.5     jsonlite_1.7.2   fs_1.5.0        
## [33] hms_1.1.0        digest_0.6.27    stringi_1.6.2    xaringan_0.21   
## [37] grid_4.1.0       rprojroot_2.0.2  cli_2.5.0        tools_4.1.0     
## [41] magrittr_2.0.1   crayon_1.4.1     pkgconfig_2.0.3  ellipsis_0.3.2  
## [45] xml2_1.3.2       reprex_2.0.0     lubridate_1.7.10 rstudioapi_0.13 
## [49] assertthat_0.2.1 rmarkdown_2.8    httr_1.4.2       R6_2.5.0        
## [53] compiler_4.1.0
```
]
---

# Where can they be found?

- CRAN &amp;ndash; official R repository
- Bioconductor &amp;ndash; mainly bioinformatics
- Github

An R package could really be uploaded anywhere, but at least CRAN and Bioconductor don't let anything through.

---

# Installing packages

.left-column[
## From CRAN
]

.right-column[

```r
# Install from CRAN
install.packages("ggplot2")
```
]

---

# Installing packages

.left-column[
## From CRAN
## From Bioconductor
]

.right-column[

```r
# Install the BiocManager from CRAN 
install.packages("BiocManager")
# Install package from Bioconductor
BiocManager::install("qvalue")
```
]

---

# Installing packages

.left-column[
## From CRAN
## From Bioconductor
## From Github
]

.right-column[

```r
# We need the devtools package to install from Github
install.packages("devtools")
# Install the development version of ggplot2
devtools::install_github("tidyverse/ggplot2")
```
]

???

BiocManager we only have to install once.
After that we will have its `install` function available.

---

# Order of operations

- Installing a package is only required once, with a few exceptions
  - The R installation is updated
  - You want to use features in a package that has been updated
- `library` needs to be called for every new session

---

# Conflicting names




```r
library(dplyr)
```

```
## 
## Attaching package: 'dplyr'
```

```
## The following object is masked from 'package:ggplot2':
## 
##     vars
```

```
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
```

---

# Conflicting names

The function that was imported last takes precedence. To access other functions, use `&lt;package&gt;::&lt;function&gt;`.


```r
intersect
```

```
## function (x, y, ...) 
## UseMethod("intersect")
## &lt;bytecode: 0x7f820f75ea00&gt;
## &lt;environment: namespace:generics&gt;
```

```r
base::intersect
```

```
## function (x, y) 
## {
##     y &lt;- as.vector(y)
##     unique(y[match(as.vector(x), y, 0L)])
## }
## &lt;bytecode: 0x7f821595c910&gt;
## &lt;environment: namespace:base&gt;
```

---

# Conflicting names

The conflicted package can help us avoid mistakes




```r
library(conflicted)
library(dplyr)

filter
```

```
## Error: [conflicted] `filter` found in 2 packages.
## Either pick the one you want with `::` 
## * dplyr::filter
## * stats::filter
## Or declare a preference with `conflict_prefer()`
## * conflict_prefer("filter", "dplyr")
## * conflict_prefer("filter", "stats")
```

---

# Conflicting names

We can then resolve the conflict for the rest of our session


```r
conflict_prefer("filter", "dplyr")
filter
```

```
## function (.data, ..., .preserve = FALSE) 
## {
##     UseMethod("filter")
## }
## &lt;bytecode: 0x7f8215c4e768&gt;
## &lt;environment: namespace:dplyr&gt;
```

---

# Packages I use (almost) daily

.pull-left.big[
- here
- conflicted
- dplyr
- ggplot2
- purrr
- stringr
- readr
]

.pull-right[
&lt;div style="position: relative; height: 280px; width: 450px;"&gt;
&lt;img style="width: 150px; margin: 0; position: absolute; top: 0; left: 0;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/ggplot2.png"&gt;
&lt;img style="width: 150px; margin: 0; position: absolute; top: 130px; left: 75px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/purrr.png"&gt;
&lt;img style="width: 150px; margin: 0; position: absolute; top: 0; left: 150px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/stringr.png"&gt;
&lt;img style="width: 150px; margin: 0; position: absolute; top: 130px; left: 225px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/dplyr.png"&gt;
&lt;img style="width: 150px; margin: 0; position: absolute; top: 0; left: 300px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/readr.png"&gt;
&lt;/div&gt;
]

---

# You can write your own R packages!

.pull-left[
RStudio is an excellent tool for developing R packages, and helps out with a lot of the formalities that need to be in place.

Not a part of this course though, but check out the book [R Packages](https://r-pkgs.org/) if you are interested.
]

.pull-right[
![:scale 75%](https://d33wubrfki0l68.cloudfront.net/19c4a5cab01d9bcb1d2edeb63ce5ba0f21870e33/68feb/images/cover.png)
]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="js/macros.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
