---
title: Practical Skills for Biology Research II
subtitle: R packages
author: Niklas Mähler
institute: "Umeå University"
date: 2021-10-04
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
source(here::here("scripts/setup.R"))
setup_presentation("r_packages")
```

# Learning goals

- Can describe what an R package is
- Know where to search for existing R packages
- Install packages from different sources
- See what packages are loaded in a session
- Basic knowledge of namespaces

---

# R packages

.left-column[
## What are they?
]

.right-column[
Source code that someone else wrote.

Collection of functions and/or data that collectively are aimed at solving a particular problem.

```{r, echo = FALSE}
tribble(
  ~Package,   ~Description,
  "ggplot2",  "Data visualisation",
  "pheatmap", "Plotting pretty heatmaps",
  "readxl",   "Reading Excel sheets into R",
  "dplyr",    "Data frame manipulation",
  "DESeq2",   "Differential gene expression analysis"
) %>% knitr::kable()
```

]

---

# R packages

.left-column[
## What are they?
## How to use them
]

.right-column[
The `library` function loads the package and makes anything exported from the package directly available.
Whatever is exported from the package will be available for the duration of the session.

```{r echo = TRUE}
library(ggplot2)
head(luv_colours)
```

> "Unpack the box and put all the items on the table"
]

---

# R packages

.left-column[
## What are they?
## How to use them
]

.right-column[
Another way of using them:

```{r echo = TRUE}
head(ggplot2::luv_colours)
```

> "Only take a specific item out of the box, use it, then put it back when you're done with it"
]

---

# R packages

.left-column[
## What are they?
## How to use them
## What is loaded?
]

.right-column[
```{r}
sessionInfo()
```
]
---

# Where can they be found?

- CRAN &ndash; official R repository
- Bioconductor &ndash; mainly bioinformatics
- Github

An R package could really be uploaded anywhere, but at least CRAN and Bioconductor don't let anything through.

---

# Installing packages

.left-column[
## From CRAN
]

.right-column[
```{r, eval = FALSE, echo = TRUE}
# Install from CRAN
install.packages("ggplot2")
```
]

---

# Installing packages

.left-column[
## From CRAN
## From Bioconductor
]

.right-column[
```{r, eval = FALSE}
# Install the BiocManager from CRAN 
install.packages("BiocManager")
# Install package from Bioconductor
BiocManager::install("qvalue")
```
]

---

# Installing packages

.left-column[
## From CRAN
## From Bioconductor
## From Github
]

.right-column[
```{r, eval = FALSE}
# We need the devtools package to install from Github
install.packages("devtools")
# Install the development version of ggplot2
devtools::install_github("tidyverse/ggplot2")
```
]

???

BiocManager we only have to install once.
After that we will have its `install` function available.

---

# Order of operations

- Installing a package is only required once, with a few exceptions
  - The R installation is updated
  - You want to use features in a package that has been updated
- `library` needs to be called for every new session

---

# Conflicting names

```{r, echo = FALSE}
# Need to clean up a bit for the example to work
detach("package:tidyverse", unload = TRUE)
unloadNamespace("modelr")
unloadNamespace("broom")
unloadNamespace("dbplyr")
detach("package:tidyr", unload = TRUE)
detach("package:dplyr", unload = TRUE)
detach("package:conflicted", unload = TRUE)
```

```{r, message = TRUE, warning = TRUE}
library(dplyr)
```

---

# Conflicting names

The function that was imported last takes precedence. To access other functions, use `<package>::<function>`.

```{r, echo = TRUE}
intersect
base::intersect
```

---

# Conflicting names

The conflicted package can help us avoid mistakes

```{r, echo = FALSE}
# More cleanup
unloadNamespace("modelr")
unloadNamespace("broom")
unloadNamespace("dbplyr")
detach("package:dplyr", unload = TRUE)
library(stats)
```

```{r, echo = TRUE, error = TRUE}
library(conflicted)
library(dplyr)

filter
```

---

# Conflicting names

We can then resolve the conflict for the rest of our session

```{r}
conflict_prefer("filter", "dplyr")
filter
```

---

# Packages I use (almost) daily

.pull-left.big[
- here
- conflicted
- dplyr
- ggplot2
- purrr
- stringr
- readr
]

.pull-right[
<div style="position: relative; height: 280px; width: 450px;">
<img style="width: 150px; margin: 0; position: absolute; top: 0; left: 0;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/ggplot2.png">
<img style="width: 150px; margin: 0; position: absolute; top: 130px; left: 75px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/purrr.png">
<img style="width: 150px; margin: 0; position: absolute; top: 0; left: 150px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/stringr.png">
<img style="width: 150px; margin: 0; position: absolute; top: 130px; left: 225px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/dplyr.png">
<img style="width: 150px; margin: 0; position: absolute; top: 0; left: 300px;" src="https://raw.githubusercontent.com/rstudio/hex-stickers/master/PNG/readr.png">
</div>
]

---

# You can write your own R packages!

.pull-left[
RStudio is an excellent tool for developing R packages, and helps out with a lot of the formalities that need to be in place.

Not a part of this course though, but check out the book [R Packages](https://r-pkgs.org/) if you are interested.
]

.pull-right[
![:scale 75%](https://d33wubrfki0l68.cloudfront.net/19c4a5cab01d9bcb1d2edeb63ce5ba0f21870e33/68feb/images/cover.png)
]