<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Journal guidelines for authors</title>
    <meta charset="utf-8" />
    <meta name="author" content="Niklas Mähler" />
    <meta name="date" content="2021-10-22" />
    <script src="libs/header-attrs-2.11/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/niklas.css" type="text/css" />
    <link rel="stylesheet" href="css/niklas-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: left, middle, inverse, title-slide

# Journal guidelines for authors
## Practical Skills for Biology Research II
### Niklas Mähler
### Umeå University
### 2021-10-22

---


layout: true

<div class="footer">
<a href="https://github.com/maehler/psbr2/tree/main/lectures/23_journal_guidelines_for_authors.Rmd">
<svg viewBox="0 0 496 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg">
<path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"></path>
</svg>
</a>
</div>

---

# From exploration to publication

The way from plotting your first figure related to a certain dataset to a figure ready for publication is long.

I will show you some snapshots of the process of making a figure for the paper titled [A single gene underlies the dynamic evolution of poplar sex determination](https://doi.org/10.1038/s41477-020-0672-9).

The basic idea was that I would map reads from sRNA to a region in the genome of *Populus tremula* in order to support the hypothesis that this region produced sRNAs that in turn silenced a gene further upstream which results in sex-determination in these trees.

---

![](../img/arr17_fig1_progress1.png)

???

I started out with this.

---

.center[
![](../img/arr17_fig1_progress2.png)
]

---

.center[
![](../img/arr17_fig1_progress3.png)
]

---

.center[
![:scale 50%](../img/arr17_fig1_progress4.png)
]

---

.center[
![](../img/arr17_fig1.png)
.right[[Müller et al. 2020](https://doi.org/10.1038/s41477-020-0672-9)]
]

???

In the end, this is what made it into the paper.

---

# Let's look at some guidelines

We'll look at the author guidelines for a couple different journals to see what they say about

- Figure formatting
- Statistical analyses
- Data sharing
- Code sharing

## Guidelines

- [New Phytologist](https://nph.onlinelibrary.wiley.com/hub/journal/14698137/about/author-guidelines)
- [eLife](https://reviewer.elifesciences.org/author-guide/full)
- [PLOS ONE](https://journals.plos.org/plosone/s/submission-guidelines)
- [F1000 Research](https://f1000research.com/for-authors/article-guidelines/research-articles)

---

# Fully reproducible articles

The publisher eLife has made an effort to allow for researchers to publish fully reproducible articles.

.https://www.nature.com/articles/nmeth.2837center[
![:scale 65%, eLife Executable Research Articles](https://iiif.elifesciences.org/journal-cms/labs-post-preview%2F2020-08%2Fimage2.jpg/full/1172,/0/default.webp)
]

So far there are only a few articles of this kind published, but I hope (and think) that this will be an integral part of publishing in the future.

???

Even though this platform was developed by eLife and Stencila, they made it open-source and available for other publishers to use.

Let's look at an example paper using this system: https://elifesciences.org/articles/30274/executable

---

# Image editing

.pull-left[
More and more journals require authors to submit raw images of gels and such.

![](https://pbs.twimg.com/profile_banners/2154127088/1623184145/1500x500)

<div class="article" style="max-width: 100;">
<time datetime="2016-04-19">2016-04-19</time>
<h1>One in 25 papers contains inappropriately duplicated images, study finds</h1>
<p class="lead">Elisabeth Bik, a microbiologist at Stanford, has for years been a behind-the-scenes force in scientific integrity, anonymously submitting reports on plagiarism and image duplication to journal editors. Now, she’s ready to come out of the shadows.</p>
<p class="source">
<a href="https://retractionwatch.com/2016/04/19/one-in-25-papers-contains-inappropriately-duplicated-images-screen-finds/">Retraction Watch</a>
</p>
</div>
]

.pull-right[
&lt;blockquote class="twitter-tweet"&gt;&lt;p lang="en" dir="ltr"&gt;Looks like someone got really creative with their gel image. &lt;a href="https://t.co/g9oxOXzJhH"&gt;pic.twitter.com/g9oxOXzJhH&lt;/a&gt;&lt;/p&gt;&amp;mdash; Elisabeth Bik (@MicrobiomDigest) &lt;a href="https://twitter.com/MicrobiomDigest/status/901677063161004032?ref_src=twsrc%5Etfw"&gt;August 27, 2017&lt;/a&gt;&lt;/blockquote&gt; &lt;script async src="https://platform.twitter.com/widgets.js" charset="utf-8"&gt;&lt;/script&gt;
]

---

# Conclusions

- Reproducibility is becoming of greater and greater concern, and more and more journals are becoming conscious about this.
- Carefully read the author guidelines for the journal you want to submit your manuscript to.
- Think about things like data sharing and transparent analyses already at the start of the project.

---

![Diagram of how data sharing ofen works today; an annoying process](../img/embl_data_sharing_bad.png)
.right[[Tony Burdett: "Open access data sharing and submission"](https://www.youtube.com/watch?v=VF12vht2qmE)]

---

![Diagram of how data sharing should look for minimal effort](../img/embl_data_sharing_good.png)
.right[[Tony Burdett: "Open access data sharing and submission"](https://www.youtube.com/watch?v=VF12vht2qmE)]

---

# Suggested reading

## On publishing

- [Introducing eLife's first computationally reproducible article](https://elifesciences.org/labs/ad58f08d/introducing-elife-s-first-computationally-reproducible-article)
  - [Replication Study: Transcriptional amplification in tumor cells with elevated c-Myc](https://elifesciences.org/articles/30274/executable)
- [Welcome to a new ERA of reproducible publishing](https://elifesciences.org/labs/dc5acbde/welcome-to-a-new-era-of-reproducible-publishing)

## On plotting

- [Visualizing samples with box plots](https://www.nature.com/articles/nmeth.2813)
  - [BoxPlotR](http://shiny.chemgrid.org/boxplotr/)
- [Kick the bar chart habit](https://www.nature.com/articles/nmeth.2837)
- [Bar charts and box plots](https://www.nature.com/articles/nmeth.2807)
- [Show the dots in plots](https://www.nature.com/articles/s41551-017-0079)

## On statistics

- [It's time to talk about ditching statistical significance](https://www.nature.com/articles/d41586-019-00874-8)
- [Scientists rise up against statistical significance](https://www.nature.com/articles/d41586-019-00857-9)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="js/macros.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"navigation": {
"scroll": false
}
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
